app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/LocalFileTransferActivity.kt

```diff
@@ -0,0 +1,382 @@
+import org.kiwix.kiwixmobile.localFileTransfer.adapter.WifiP2pDelegate
+import org.kiwix.kiwixmobile.localFileTransfer.adapter.WifiPeerListAdapter
+import java.util.ArrayList
+import javax.inject.Inject
+
+/**
+ * Created by @Aditya-Sood as a part of GSoC 2019.
+ *
+ * This activity is the starting point for the module used for sharing zims between devices.
+ *
+ * The module is used for transferring ZIM files from one device to another, from within the
+ * app. Two devices are connected to each other using WiFi Direct, followed by file transfer.
+ *
+ * File transfer involves two phases:
+ * 1) Handshake with the selected peer device, using [PeerGroupHandshakeAsyncTask]
+ * 2) After handshake, starting the files transfer using [SenderDeviceAsyncTask] on the sender
+ * device and [ReceiverDeviceAsyncTask] files receiving device
+ */
+@SuppressLint("GoogleAppIndexingApiWarning", "Registered")
+class LocalFileTransferActivity : BaseActivity(),
+  WifiDirectManager.Callbacks {
+  @Inject
+  lateinit var alertDialogShower: AlertDialogShower
+
+  @Inject
+  lateinit var wifiDirectManager: WifiDirectManager
+
+  @Inject
+  lateinit var locationManager: LocationManager
+
+  private var isFileSender = false // Whether the device is the file sender or not
+  private var filesForTransfer = ArrayList<FileItem>()
+  private var fileListAdapter: FileListAdapter? = null
+  private var wifiPeerListAdapter: WifiPeerListAdapter? = null
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_local_file_transfer)
+
+    /*
+     * Presence of file Uris decides whether the device with the activity open is a sender or receiver:
+     * - On the sender device, this activity is started from the app chooser post selection
+     * of files to share in the Library
+     * - On the receiver device, the activity is started directly from within the 'Get Content'
+     * activity, without any file Uris
+     * */
+    val filesIntent = intent
+    val fileUriArrayList: ArrayList<Uri>?
+    fileUriArrayList = filesIntent.getParcelableArrayListExtra(Intent.EXTRA_STREAM)
+    isFileSender = fileUriArrayList?.isNotEmpty() == true && fileUriArrayList.size > 0
+
+    val toolbar: Toolbar =
+      findViewById(R.id.toolbar)
+    // I know this is not the best method, but android extensions were showing 'unknown reference' error
+    setSupportActionBar(toolbar)
+    toolbar.setNavigationIcon(R.drawable.ic_close_white_24dp)
+    toolbar.setNavigationOnClickListener { finish() }
+    wifiPeerListAdapter = WifiPeerListAdapter(
+      WifiP2pDelegate(wifiDirectManager::sendToDevice)
+    )
+
+    list_peer_devices.adapter = wifiPeerListAdapter
+    list_peer_devices.layoutManager = LinearLayoutManager(this)
+    list_peer_devices.setHasFixedSize(true)
+    if (isFileSender) {
+      fileUriArrayList.map {
+        filesForTransfer.add(FileItem(it))
```

Comment:
map produces a new list, here is an example scratch

![image](https://user-images.githubusercontent.com/3358282/90240003-6307d980-de20-11ea-90fe-cd9057a914c4.png)

Reply:
What can I do to solve this issue?

Reply:
What I feel is that they were trying to tell you that the `map` _creates a new list_ and we already have one (the one that you've previously named as `fileUriArrayList`), so probably you're being told to avoid that overkill and find out some other alternative to implement the same :)

Reply:
What you're trying to say is, I should use some other function instead of `map` to solve this issue, or use `map` function but in a different way?

Reply:
Here you are using `map` as `forEach` when the expression of this code should be `transform uris into FileItems`

Say I had a list of names and wanted to create a list of some Person objects

```
data class Person(val name:String, val age:Int)

val listOfPeople = listOf("Alice","Bob","Ted", "Carol").map{ Person(it, 0..100.random())}
```

Does it make more sense now?

Reply:
Just to be clear, in this example,  `Person class` is synonymous to `FileItem class` and `listOf("Alice","Bob","Ted", "Carol")` is to `fileUriArrayList` and `listOfPeople` is to filesForTransfer  ?

Reply:
`filesForTransfer = fileUriArrayList.map { FileItem(it) }` 

would this suffice ?

Reply:
That is what I wanted you take from my example.

Reply:
Since `filesForTransfer` is an ArrayList and `map` produces List, I used this piece of code to solve the issue 

`filesForTransfer = ArrayList(fileUriArrayList.map(::FileItem))`

any comments ?

Reply:
change the type of `filesForTransfer`, explicitly mentioning the implementation of `List` you are using is not great practice